# Project: Analyzing borrowers’ risk of defaulting


# Context:
This project is about a report for a Bank's loan division. The management wants  to find out if the marital status of a customer and number of children has an impact on whether they will default on a loan or not. The bank already has some data on customers’ credit worthiness. This report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan. 
I would be testing the following Hypotheses: 
- Is there is a correlation between having children and paing back on time? 
- Is there a correlation between family status and paying back on time?
- Is there a correlation between Income level and paying back on time?
- How does Credit purpose affect the default rate?

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan


## Reading the dataset 

In [1]:
# Loading all the libraries
import pandas as pd

# Load the data
df = pd.read_csv("/datasets/credit_scoring_eng.csv")



## Data exploration

In [2]:
# Let's see how many rows and columns our dataset has
df.shape

(21525, 12)

In [3]:
# Print the first ten(10) rows
df.head(10)


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

From the table, the first ten (10) rows were printed. The days_employed column has both float and negative values. Those values have to be changed to integers, and the negative values have to be removed. 

In [4]:
# Get info on data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


There are missing values in two columns: days_employed columns and total_income column.They have 19351 values, while other columns have a total of 21525 values.

In [5]:
# Let's look in the filtered table at the first column with missing data
print(df["days_employed"].isnull().sum())


2174


In [6]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.
df[(df["total_income"].isna()) & (df["days_employed"].isna())] #& function looks at the two columns for missing values simutaneously

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

The missing values in both columns are symmetric. The number of missing values in total_income matches the number of missing values in days_employed. 


In [7]:
#The percentage of missing values in the whole dataset

missing_value = 2174
total_value = 21525
percent_missing_value = missing_value/total_value

print(f"The percentage of missing values is: {percent_missing_value: .0%}")

The percentage of missing values is:  10%


**Intermediate conclusion**
The number of rows in the filtered table match the number of missing values. There are 2174 missing values in both tables

The percentage of the missing values in relation to the whole dataset is 10%. The dataset is not large. So, I will be filling in the missing values with either the median or the mean based on the client characteristic that gives a considerable variation in distribution of the mean, and median, and not the global mean or median of the days_employed column and total_income column.

Subsequently, I would compare the distribution of the data between the missing values for identified client characteristics and the whole dataset to see patterns in the data. Also, I would take out duplicates, remove negative values, and fill in the missing values in the identified columns.

In [8]:
df.head(50)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   
15         1    -972.364419         26  secondary education             1   
16         0   -1719.934226         35  secondary education             1   
17         0   -2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0  -10038.818549         48  SECONDARY EDUCATION             1   
20         1   -1311.604166         36  secondary education             1   
21         1    -253.685166         33  secondary education             1   
22         1   -1766.644138         24  secondary education             1   
23         0    -272.981385         21    bachelor's degree             0   
24         1  338551.952911         57  secondary education             1   
25         0  363548.489348         67  secondary education             1   
26         0            NaN         41  secondary education             1   
27         0    -529.191635         28    bachelor's degree             0   
28         1    -717.274324         26    bachelor's degree             0   
29         0            NaN         63  secondary education             1   
30         1  335581.668515         62  secondary education             1   
31         0   -1682.083438         47    primary education             3   
32         0   -4649.910832         34  SECONDARY EDUCATION             1   
33         0   -1548.637544         48  secondary education             1   
34         0   -4488.067031         35  secondary education             1   
35         0  394021.072184         68  secondary education             1   
36         0    -176.216688         33  secondary education             1   
37         0   -6448.810860         43    BACHELOR'S DEGREE             0   
38         0    -597.881827         25    bachelor's degree             0   
39         0    -650.587796         31    bachelor's degree             0   
40         0   -1030.362606         30  secondary education             1   
41         0            NaN         50  secondary education             1   
42         0   -1257.496190         20         some college             2   
43         0   -4375.681384         43         some college             2   
44         1   -1362.041728         26  SECONDARY EDUCATION             1   
45         1   -1039.451741         49  secondary education             1   
46         2   -2262.712304         37  secondary education             1   
47         1   -2689.137274         33    bachelor's degree             0   
48         0   -3341.067886         45  secondary education             1   
49         0   -1181.443228         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
0    

In [9]:
#Investigate clients who do not have data on identified characteristics and the column with the missing values
missing_data = df[(df["days_employed"].isna()) | (df["total_income"].isna())]
missing_data["children"].value_counts()
# comparing distribution of null value dataframe to entire dataframe on children column

 0     1439
 1      475
 2      204
 3       36
 20       9
 4        7
-1        3
 5        1
Name: children, dtype: int64

In [10]:
#comparing distribution of non-null value dataframe to entire dataframe on children column
missing_data["children"].value_counts()/missing_data.shape[0]

 0     0.661914
 1     0.218491
 2     0.093836
 3     0.016559
 20    0.004140
 4     0.003220
-1     0.001380
 5     0.000460
Name: children, dtype: float64

In [11]:
#comparing distribution of non-null value dataframe to entire dataframe on children column
df["children"].value_counts()/df.shape[0]


 0     0.657329
 1     0.223833
 2     0.095470
 3     0.015331
 20    0.003531
-1     0.002184
 4     0.001905
 5     0.000418
Name: children, dtype: float64

The distribution of the filtered dataset and the whole dataset on the children columnn is similar. Also there are outlier values of -1 and 20 that are present in the filtered set whhich will be handled later.

In [12]:
#comparing distribution of non-null value dataframe to entire dataframe on days_employed column
missing_data["education"].value_counts()/missing_data.shape[0]


secondary education    0.647654
bachelor's degree      0.228151
SECONDARY EDUCATION    0.030819
Secondary Education    0.029899
some college           0.025299
Bachelor's Degree      0.011500
BACHELOR'S DEGREE      0.010580
primary education      0.008740
Some College           0.003220
SOME COLLEGE           0.003220
Primary Education      0.000460
PRIMARY EDUCATION      0.000460
Name: education, dtype: float64

In [13]:
#comparing distribution of non-null value dataframe to entire dataframe on children column
df["education"].value_counts()/df.shape[0]


secondary education    0.638792
bachelor's degree      0.219187
SECONDARY EDUCATION    0.035865
Secondary Education    0.033031
some college           0.031034
BACHELOR'S DEGREE      0.012729
Bachelor's Degree      0.012451
primary education      0.011614
Some College           0.002184
SOME COLLEGE           0.001347
PRIMARY EDUCATION      0.000790
Primary Education      0.000697
graduate degree        0.000186
Graduate Degree        0.000046
GRADUATE DEGREE        0.000046
Name: education, dtype: float64

There are values for rows that are present in the non-null dataframe, and are absent in the entire dataframe. After taking a close look, I found out that they are duplicates. Those duplicates would be dealt with later in the course of this project. 
Also, it was discovered that there is graduate degree in the entire dataframe, which is missing in the filtered dataset.
The distribution of this columnn is similar between the filtered dataset and the whole dataset.

In [14]:
#comparing distribution of non-null value dataframe to entire dataframe on income_type column
missing_data["income_type"].value_counts()/missing_data.shape[0]


employee         0.508280
business         0.233671
retiree          0.189972
civil servant    0.067617
entrepreneur     0.000460
Name: income_type, dtype: float64

In [15]:
#comparing distribution of non-null value dataframe to entire dataframe on income_type column
df["income_type"].value_counts()/df.shape[0]


employee                       0.516562
business                       0.236237
retiree                        0.179141
civil servant                  0.067782
unemployed                     0.000093
entrepreneur                   0.000093
student                        0.000046
paternity / maternity leave    0.000046
Name: income_type, dtype: float64

The income_type column has two more category of people, which are students, and those on paternity/maternity leave in the whole dataset. The distribution for this columnn is similar between the rows of data that have null values, and the rows of data that do not have null values. 


In [16]:
#comparing distribution of non-null value dataframe to entire dataframe on family_status column
missing_data["family_status"].value_counts()/missing_data.shape[0]


married              0.568997
civil partnership    0.203312
unmarried            0.132475
divorced             0.051518
widow / widower      0.043698
Name: family_status, dtype: float64

In [17]:
#comparing distribution of non-null value dataframe to entire dataframe on income_type column
df["family_status"].value_counts()/df.shape[0]

married              0.575145
civil partnership    0.194053
unmarried            0.130685
divorced             0.055517
widow / widower      0.044599
Name: family_status, dtype: float64

The distribution of the family_status columnn is similar between the filtered dataset, and the whole dataset. 

**Possible reasons for missing values in data**

The missing values in our dataset are given as Nan(Not a Number), which is a float, telling us that mathematical operations can be performed on them. The missing values in this dataset are most likely due to errors when copying data, or technical errors. 


**Intermediate conclusion**
The distribution in the original dataset is very similar to the filtered table. Therefore, we can conclude that missing values in the dataset are randomly generated. There is no specific pattern to them


**Conclusions**

No patterns were found in the dataset. This conclusion was arrived at due to the similarity in the distribution of the filtered dataset, and the whole dataset.

To transform the data, I found out that there are duplicates in the education column. I have to take those duplicates out. Also for the sake of ease in manipulating data, I have to convert all the values in total_income column from floats to integers. The negative values in the days_employed column would be removed by taking the absolute value. Also, outlier values will be removed from the column by taking out all the years greater than 60, and converting them to days. Lastly, I would replace the missing values in the total_income column with the median.

## Data transformation



In [18]:
# Values in education column 
print(df["education"].value_counts())

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64


In [19]:
# After errors were fixed
df["education"] = df["education"].str.lower() #applies the lower string to all the rows in the education column
df["education"].value_counts() #

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

All the values in the "education" column have been converted to lowercase, and the incorrect registers have been fixed, thereby eliminating duplicates

In [20]:
# Checking the distribution in the "children's column"
df["children"].value_counts()/df["children"].shape[0]

 0     0.657329
 1     0.223833
 2     0.095470
 3     0.015331
 20    0.003531
-1     0.002184
 4     0.001905
 5     0.000418
Name: children, dtype: float64

In [21]:
neg_values_children = df.loc[df["children"] == -1].shape[0]
outlier_values_children = df.loc[df["children"] == 20].shape[0]

sum_total = (neg_values_children + outlier_values_children)/df.shape[0]

print(f"The number of negative and outlier values is {sum_total}")

The number of negative and outlier values is 0.005714285714285714


There are values of -1, and 20 in the children's column, these errors make up 0.006% of the whole dataset. These errors could have stemmed from mistakes while filling in the data. Also, the outlier values might be due totechnical errors while transferring the data. To address this issue, the median of the whole children column is calculated, and it is used to replace the -1 and 20 values

In [22]:
# To fix the data 

median = df.loc[(df['children'] != -1) & (df['children'] != 20), 'children'].median()

#df.loc[(df['children'] == -1) or (df['children'] == 20), "children"] = median

df.loc[(df['children'] == -1) | (df['children'] == 20), "children"] = median

In [23]:
df["children"].value_counts()

0.0    14272
1.0     4818
2.0     2055
3.0      330
4.0       41
5.0        9
Name: children, dtype: int64

In the days_employed column the data has negative values, also there are outliers in the dataset. To fix this column, I would take the absolute value of the days_employed column to remove the negative values.

In [24]:
df[df["days_employed"] < 0].shape[0]/df.shape[0]

0.7389547038327526

The number of values that are negative in the days_employed column is 15906, and the percentage is 73.8%

In [25]:
df.loc[df["days_employed"] < 0, "days_employed"].shape[0]/df.shape[0] #gives the percentage of the negative values in the dataset

0.7389547038327526

In [26]:
#First we need to get the absolute value of days_employed column to remove the -ve values
df["days_employed"] = df["days_employed"].abs()

In [27]:
#Getting the maximum value from days_employed column
df["days_employed"].max()

401755.40047533

In [28]:
#Getting the minimum value from days_employed column
df["days_employed"].min()

24.14163324048118

The maximum number of days worked by a person in the day_employed column is 401,755. When converted to years, that is 1,100 years. This is obviously an outlier value. To take out the outlier values, We filter out all ages greater than 65 years. This is because 65 is the average age for retirement.

In [29]:
#To remove outlier values, I calculate the number of years worked into days 65years * 365 days =23725
df.loc[df["days_employed"] > 23725, "days_employed"] = df.loc[df["days_employed"] < 23725, "days_employed"].median()

In [30]:
df["days_employed"].value_counts() #outliers have been replace with the median. Also, 3445 is the number of the outliers.

1630.019381    3445
142.276217        1
144.185854        1
2569.204627       1
3545.955468       1
               ... 
2152.475526       1
6219.157426       1
5873.586442       1
3355.260838       1
1636.419775       1
Name: days_employed, Length: 15907, dtype: int64

The outlier values have been replaced with the median age. The number of values that are outliers is 3445

Taking a look at the client's age to se if there are any anomalies in the data. 

In [31]:
# Check the `dob_years` for suspicious values and count the percentage
df["dob_years"].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Of the 21525 values in the dob_years columns, there are 101 people who have their date of birth recorded as 0. We fill these rows with mean of the dob_years column

In [32]:
# Address the issues in the `dob_years` column, if they exist
df.loc[df["dob_years"] == 0, "dob_years"] = df.loc[df["dob_years"] > 0, "dob_years"].mean()

In [33]:
# Check the result - make sure it's fixed
df[df["dob_years"] == 0].shape[0]

0

Checking the values for the family_status column, to see if there are any issues.

In [34]:
# Let's see the values for the column
df["family_status"].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

The family_status column has no issues. 



Taking a look at the gender column to fix the issues there.

In [35]:
# Let's see the values in the column
df["gender"].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

There are about 14236 Females, and 7288 Males. However, there is a strange gender XNA. I will leave it like that.

Checking the income_type column

In [36]:
# Let's see the values in the column
df["income_type"].value_counts()

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

There are no issues with the income_type column.

 Next, I check if there are any duplicates in my data. If there are, I take them out.

In [37]:
# Checking duplicates
df.duplicated().sum()


71

In [38]:
# Addressing the duplicates, if they exist
df = df.drop_duplicates().reset_index(drop=True)

In [39]:
# Checking again for duplicates
df.duplicated().sum()

0

In [40]:
# Check the size of the dataset that you now have after your first manipulations with it
df.shape

(21454, 12)

After taking out the duplicate values from the dataset, there are 21454 rows of data in the dataset.


# Working with missing values

To speed up working with our dataset, I will pair some columns that form a dictionary namely: education, education id, and family, family_status_id. This was done because the columns are closely related, and they give information about each other.

In [41]:
# Pairing the dictionaries education and education_id
df[["education","education_id"]].drop_duplicates().reset_index(drop=True)


education  education_id
0    bachelor's degree             0
1  secondary education             1
2         some college             2
3    primary education             3
4      graduate degree             4

In [42]:
# Pairing the dictionaries family_status, and family_status_id
df[["family_status", "family_status_id"]].drop_duplicates().reset_index(drop=True)

family_status  family_status_id
0            married                 0
1  civil partnership                 1
2    widow / widower                 2
3           divorced                 3
4          unmarried                 4

### Restoring missing values in `total_income`

The days_column, and total_income, both have missing values. Here I would deal with the total_income, by checking the missing values, and replcaing with either the mean or median. First, I would check for outliers. If there are outliers, then I would use the median. If there are no outliers, then I would use the median. 
However, I would first group the ages into categories, to enable me calculate the values for the total income.


In [43]:
# Let's write a function that calculates the age category
def age_category(age):
    if age < 18:
        return "below 18"
    if age > 18 and age < 25:
        return "18 - 25"
    if age >25 and age < 45:
        return "25 - 45"
    if age > 45 and age < 65:
        return "45 -65"
    return "above 65" 

    

In [44]:
#Checking the Max age
df["dob_years"].max()

75.0

In [45]:
#Checking the minimum age
df["dob_years"].min()

19.0

In [46]:
# Testing the function
print(age_category(20))

18 - 25


In [47]:
# Creating new column based on function
df["dob_years_groups"] = df["dob_years"].apply(age_category)

In [48]:
# Checking how values in the new column
df

children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028       42.0    bachelor's degree             0   
1           1.0    4024.803754       36.0  secondary education             1   
2           0.0    5623.422610       33.0  secondary education             1   
3           3.0    4124.747207       32.0  secondary education             1   
4           0.0    1630.019381       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21449       1.0    4529.316663       43.0  secondary education             1   
21450       0.0    1630.019381       67.0  secondary education             1   
21451       1.0    2113.346888       38.0  secondary education             1   
21452       3.0    3112.481705       38.0  secondary education             1   
21453       2.0    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21449  civil partnership                 1      F    business     0   
21450            married                 0      F     retiree     0   
21451  civil partnership                 1      M    employee     1   
21452            married                 0      M    employee     1   
21453            married                 0      F    employee     0   

       total_income                  purpose dob_years_groups  
0         40620.102    purchase of the house          25 - 45  
1         17932.802             car purchase          25 - 45  
2         23341.752    purchase of the house          25 - 45  
3         42820.568  supplementary education          25 - 45  
4         25378.572        to have a wedding           45 -65  
...             ...                      ...              ...  
21449     35966.698     housing transactions          25 - 45  
21450     24959.969        purchase of a car         above 65  
21451     14347.610                 property          25 - 45  
21452     39054.888        buying my own car          25 - 45  
21453     13127.587             to buy a car          25 - 45  

[21454 rows x 13 columns]

In [49]:
#Taking a look at the groups based on their grouped income 
df["dob_years_groups"].value_counts()

25 - 45     10576
45 -65       8255
above 65     1748
18 - 25       875
Name: dob_years_groups, dtype: int64

To fill the missing values in the column total_income, I would use the mean or median based on their distribution. Then look at the columns that affect the total_income to determine the column that gives the best distribution for either the mean or the median.

I would create a table without missing values. This new table will be used to fill the missing values

In [50]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
df_filled_values = df.loc[(df["days_employed"].isna()!=True) & (df["total_income"].isna()!= True)]
df_filled_values.head(5)




children  days_employed  dob_years            education  education_id  \
0       1.0    8437.673028       42.0    bachelor's degree             0   
1       1.0    4024.803754       36.0  secondary education             1   
2       0.0    5623.422610       33.0  secondary education             1   
3       3.0    4124.747207       32.0  secondary education             1   
4       0.0    1630.019381       53.0  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose dob_years_groups  
0    purchase of the house          25 - 45  
1             car purchase          25 - 45  
2    purchase of the house          25 - 45  
3  supplementary education          25 - 45  
4        to have a wedding           45 -65

In [51]:
# Look at the mean values for total_income 
df_filled_values["total_income"].mean()


26787.568354658673

In [52]:
# Look at the median values for total_income 
total_income_median = df_filled_values["total_income"].median()
total_income_median

23202.87

In [53]:
#Getting the minimum value for total income 
df_filled_values["total_income"].min()

3306.762

In [54]:
#Getting the maximum value for total income 
df_filled_values["total_income"].max()

362496.645

I would be using the median to fill in the missing values. The difference between the minimum value and maximum values are quite large, therefore, they act like outliers, and would skew the result if I were to use the mean.  


In [55]:
#Calculating the global median, and using it to fill all the missing values, and assigning it to a variable   

df["total_income_global_med"] = df["total_income"].fillna(df_filled_values["total_income"].median() ) #df_filled_values[total_income].median()


In [56]:
df_filled_values["total_income"].median()

23202.87

In [57]:
#Taking a look at the distribution of the median of the total_income based on the family_status 
df[['family_status','total_income']].groupby('family_status').median()

total_income
family_status                  
civil partnership     23186.534
divorced              23515.096
married               23389.540
unmarried             23149.028
widow / widower       20514.190

The distribution of the median of the family_status does not really affect the distribution of the total_income 

In [58]:
#Taking a look at the distribution of the median of the total_income based on the education
df[['education','total_income']].groupby('education').median()

total_income
education                        
bachelor's degree      28054.5310
graduate degree        25161.5835
primary education      18741.9760
secondary education    21836.5830
some college           25618.4640

The distribution of the median of the education shows variations in the total Income that was earned by various category of people. So, we can use the median of this group. However, let's try the third column that affects total_income, which is income_type. 

In [59]:
#Taking a look at the distribution of the median of the total_income based on the income_type 
df[['income_type','total_income']].groupby('income_type').median()

total_income
income_type                              
business                       27577.2720
civil servant                  24071.6695
employee                       22815.1035
entrepreneur                   79866.1030
paternity / maternity leave     8612.6610
retiree                        18962.3180
student                        15712.2600
unemployed                     21014.3605

The distribution of the median of the income_type gives the best variations in the total Income that was earned by various category of people. So, we the median of income_type will be used to fill the missing values. 

In [60]:
 df["total_income_gb_med"] = df[['income_type','total_income']].groupby("income_type").transform(lambda x: x.fillna(x.median()))

In [61]:
df["total_income"] = df["total_income_gb_med"]

In [62]:
df['total_income'].describe()

count     21454.000000
mean      26451.212929
std       15709.968189
min        3306.762000
25%       17219.817250
50%       22815.103500
75%       31331.348000
max      362496.645000
Name: total_income, dtype: float64

In [63]:
#Deleting the colums that are not needed
del df["total_income_gb_med"]
del df["total_income_global_med"] 

To see that the values in total_income matches the number of values in others

In [64]:
# Checking the number of entries in the columns
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  float64
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  float64
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  dob_years_groups  21454 non-null  object 
dtypes: float64(4), int64(3), object(6)
memory usage: 2.1+ MB


All missing values have been filled in the total_income column

###  Restoring values in `days_employed`

To fill the missing values in the column days_employed, I would use the mean or median based on their distribution. Then look at the columns that affect the days_employed column to determine the column that gives the best distribution for either the mean or the median.

In [65]:
#Calculating the mean for days_employed column
df_filled_values["days_employed"].mean()

2224.3030428320353

In [66]:
#Calculating the median for days_employed column
df_filled_values["days_employed"].median()

1630.0193809778218

In [67]:
#Calculating the maximum value in the days_employed columns
df_filled_values["days_employed"].max()

18388.949900568383

In [68]:
#Calculating the minimum value in the days_employed columns
df_filled_values["days_employed"].min()

24.14163324048118

The difference between the minimum value, and the maximum value in the days_employed column is quite large, and would act as ouliers, so it might be somewhat tricky to use the mean to fill the missing values. However, let's see their distribution first. Also, the parameter identified that could be used to determine the median of the days_employed column is dob_years.

In [69]:
df_filled_values

children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028       42.0    bachelor's degree             0   
1           1.0    4024.803754       36.0  secondary education             1   
2           0.0    5623.422610       33.0  secondary education             1   
3           3.0    4124.747207       32.0  secondary education             1   
4           0.0    1630.019381       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21449       1.0    4529.316663       43.0  secondary education             1   
21450       0.0    1630.019381       67.0  secondary education             1   
21451       1.0    2113.346888       38.0  secondary education             1   
21452       3.0    3112.481705       38.0  secondary education             1   
21453       2.0    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21449  civil partnership                 1      F    business     0   
21450            married                 0      F     retiree     0   
21451  civil partnership                 1      M    employee     1   
21452            married                 0      M    employee     1   
21453            married                 0      F    employee     0   

       total_income                  purpose dob_years_groups  
0         40620.102    purchase of the house          25 - 45  
1         17932.802             car purchase          25 - 45  
2         23341.752    purchase of the house          25 - 45  
3         42820.568  supplementary education          25 - 45  
4         25378.572        to have a wedding           45 -65  
...             ...                      ...              ...  
21449     35966.698     housing transactions          25 - 45  
21450     24959.969        purchase of a car         above 65  
21451     14347.610                 property          25 - 45  
21452     39054.888        buying my own car          25 - 45  
21453     13127.587             to buy a car          25 - 45  

[19351 rows x 13 columns]

In [70]:
#Distribution of `days_employed` medians based on identified parameters

df_filled_values[["dob_years","days_employed"]].groupby("dob_years").median()                

days_employed
dob_years               
19.000000     724.492610
20.000000     674.838979
21.000000     618.733817
22.000000     703.310078
23.000000     690.204208
24.000000     947.731043
25.000000     919.199388
26.000000    1083.658132
27.000000    1166.212160
28.000000    1141.705450
29.000000    1315.453550
30.000000    1420.586863
31.000000    1308.901135
32.000000    1446.622542
33.000000    1426.415003
34.000000    1615.910188
35.000000    1613.494818
36.000000    1782.652308
37.000000    1783.979003
38.000000    1738.115310
39.000000    1832.564621
40.000000    1674.726020
41.000000    1847.778632
42.000000    2143.026818
43.000000    1748.001959
43.497479    1630.019381
44.000000    1813.405616
45.000000    2105.382664
46.000000    1987.334781
47.000000    1999.975513
48.000000    2076.293896
49.000000    2138.776097
50.000000    1843.949478
51.000000    1630.019381
52.000000    1630.019381
53.000000    1630.019381
54.000000    1630.019381
55.000000    1630.019381
56.000000    1630.019381
57.000000    1630.019381
58.000000    1630.019381
59.000000    1630.019381
60.000000    1630.019381
61.000000    1630.019381
62.000000    1630.019381
63.000000    1630.019381
64.000000    1630.019381
65.000000    1630.019381
66.000000    1630.019381
67.000000    1630.019381
68.000000    1630.019381
69.000000    1630.019381
70.000000    1630.019381
71.000000    1630.019381
72.000000    1630.019381
73.000000    1630.019381
74.000000    1630.019381
75.000000    1678.969771

In [71]:
# Distribution of `days_employed` means based identified parameters

df_filled_values[["dob_years","days_employed"]].groupby("dob_years").mean()  


days_employed
dob_years               
19.000000     633.678086
20.000000     684.944308
21.000000     709.440930
22.000000     786.489080
23.000000     827.309437
24.000000    1026.405485
25.000000    1088.406453
26.000000    1202.592242
27.000000    1359.938157
28.000000    1398.193809
29.000000    1553.823200
30.000000    1696.039355
31.000000    1652.672079
32.000000    1735.111375
33.000000    1867.754671
34.000000    1981.737277
35.000000    2108.015677
36.000000    2268.854680
37.000000    2173.264194
38.000000    2297.106442
39.000000    2400.613823
40.000000    2336.063602
41.000000    2424.813669
42.000000    2773.492526
43.000000    2475.880813
43.497479    2093.825679
44.000000    2790.829304
45.000000    2762.729538
46.000000    2829.449013
47.000000    2922.221232
48.000000    2846.645423
49.000000    3000.855655
50.000000    2903.220400
51.000000    2789.821854
52.000000    2902.851802
53.000000    2863.137305
54.000000    2518.309634
55.000000    2778.809851
56.000000    2573.797527
57.000000    2488.516077
58.000000    2427.314021
59.000000    2480.580121
60.000000    2237.113634
61.000000    2411.015296
62.000000    1982.023158
63.000000    2181.190034
64.000000    2149.105824
65.000000    2160.555302
66.000000    1964.240195
67.000000    1879.669148
68.000000    1954.621705
69.000000    1790.961161
70.000000    2021.781857
71.000000    1738.023479
72.000000    2111.364608
73.000000    1887.045967
74.000000    2488.762978
75.000000    1678.969771

From the distribution of both the median and the mean, it is clearly seen that the mean gives a better distribution for the various ages. So, we use the mean to fill the missing values

In [72]:
# Replacing missing values
df["days_employed"] = df[["dob_years","days_employed"]].groupby("dob_years").transform(lambda x: x.fillna(x.mean()))


In [73]:
df["days_employed"].describe()

count    21454.000000
mean      2225.323883
std       2009.110764
min         24.141633
25%       1011.981643
50%       1630.019381
75%       2762.729538
max      18388.949901
Name: days_employed, dtype: float64

All missing values have been fixed in days_employed column

In [74]:
# Checking the entries in all columns to make sure we fixed all missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  float64
 1   days_employed     21454 non-null  float64
 2   dob_years         21454 non-null  float64
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  dob_years_groups  21454 non-null  object 
dtypes: float64(4), int64(3), object(6)
memory usage: 2.1+ MB


## Categorization of data

To answer the questions and test the hypotheses, I will have to categorize the data. Of all the identified characteristics, I will categorize the data in the "purpose" column. 
[Despite of how you decide to address the categorization, make sure to provide clear explanation of why you made your decision. Remember: this is your work and you make all decisions in it.]


In [75]:
# Print the values for your selected data for categorization
df["purpose"].value_counts()

wedding ceremony                            791
having a wedding                            768
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
housing                                     646
purchase of the house                       646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             621
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

[Let's check unique values]

In [76]:
# Check the unique values
df["purpose"].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

The main groups that I can identify here are groups with Cars, Education, House, Wedding


In [77]:
# Here I create a function to assign several purposes for loans as a single parameter

input_text = "wedding ceremony"

def text_normalizer(text):
    if 'car' in text:
        return 'cars'
    elif ( 'hous' in text ) or  ( 'real estate' in text ) or  ( 'property' in text ):
        return 'real estate'
    elif  'wedding' in text :
        return 'wedding'
    elif ( 'educat' in text ) or  ( 'university' in text ):
        return 'education'
    else:
        return 'no category'
    

In [78]:
text_normalizer(input_text)

'wedding'

In [79]:
df["purpose_norm"] = df["purpose"].apply(lambda x:text_normalizer(x) )

In [80]:
df["purpose_norm"].value_counts()

real estate    10811
cars            4306
education       4013
wedding         2324
Name: purpose_norm, dtype: int64

**Categorization of debt** 


The debt column gives just two values 0 and 1. 0 means debt paid on time, 1 means default

In [81]:
df["debt"].value_counts()

0    19713
1     1741
Name: debt, dtype: int64

In [82]:
def debt_profile(debt_sta):#this defines a function  debt_profile, and it takes the argument debt_sta
    
    
    if debt_sta == 0:
        return "paid"
    #if the input value is 0, the string returned is "paid"
    elif debt_sta == 1:
        return "defaulted"
    #if the input value is 1, the string returned is "defaulted"
    
df["debt_profile"] = df["debt"].apply(debt_profile)
df.head(10)    

children  days_employed  dob_years            education  education_id  \
0       1.0    8437.673028       42.0    bachelor's degree             0   
1       1.0    4024.803754       36.0  secondary education             1   
2       0.0    5623.422610       33.0  secondary education             1   
3       3.0    4124.747207       32.0  secondary education             1   
4       0.0    1630.019381       53.0  secondary education             1   
5       0.0     926.185831       27.0    bachelor's degree             0   
6       0.0    2879.202052       43.0    bachelor's degree             0   
7       0.0     152.779569       50.0  secondary education             1   
8       2.0    6929.865299       35.0    bachelor's degree             0   
9       0.0    2188.756445       41.0  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose dob_years_groups purpose_norm  \
0                purchase of the house          25 - 45  real estate   
1                         car purchase          25 - 45         cars   
2                purchase of the house          25 - 45  real estate   
3              supplementary education          25 - 45    education   
4                    to have a wedding           45 -65      wedding   
5                purchase of the house          25 - 45  real estate   
6                 housing transactions          25 - 45  real estate   
7                            education           45 -65    education   
8                     having a wedding          25 - 45      wedding   
9  purchase of the house for my family          25 - 45  real estate   

  debt_profile  
0         paid  
1         paid  
2         paid  
3         paid  
4         paid  
5         paid  
6         paid  
7         paid  
8         paid  
9         paid

**Categorizing the "total_income" column** 

In [83]:
df["total_income"].max()

362496.645

In [84]:
df["total_income"].min()

3306.762

In [85]:
def income_status(total_income):
#created function "income_status" with total_income as the argument

    if total_income > 0 and total_income < 40000:
        return "low earner"
    if total_income > 40000 and total_income < 80000:
        return "average earner"
    if total_income > 80000 and total_income < 120000:
        return "above average earner"
    if total_income > 120000 and total_income < 180000:
        return "High earner"
    if total_income > 180000 and total_income < 250000:
        return "Very High earner"
    return "Extremely High earner"
        
    
df["income_status"] = df["total_income"].apply(income_status) 
#Here we apply our function to the total_income column, and it creates another column "income_status" 


In [86]:
df

children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028       42.0    bachelor's degree             0   
1           1.0    4024.803754       36.0  secondary education             1   
2           0.0    5623.422610       33.0  secondary education             1   
3           3.0    4124.747207       32.0  secondary education             1   
4           0.0    1630.019381       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21449       1.0    4529.316663       43.0  secondary education             1   
21450       0.0    1630.019381       67.0  secondary education             1   
21451       1.0    2113.346888       38.0  secondary education             1   
21452       3.0    3112.481705       38.0  secondary education             1   
21453       2.0    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21449  civil partnership                 1      F    business     0   
21450            married                 0      F     retiree     0   
21451  civil partnership                 1      M    employee     1   
21452            married                 0      M    employee     1   
21453            married                 0      F    employee     0   

       total_income                  purpose dob_years_groups purpose_norm  \
0         40620.102    purchase of the house          25 - 45  real estate   
1         17932.802             car purchase          25 - 45         cars   
2         23341.752    purchase of the house          25 - 45  real estate   
3         42820.568  supplementary education          25 - 45    education   
4         25378.572        to have a wedding           45 -65      wedding   
...             ...                      ...              ...          ...   
21449     35966.698     housing transactions          25 - 45  real estate   
21450     24959.969        purchase of a car         above 65         cars   
21451     14347.610                 property          25 - 45  real estate   
21452     39054.888        buying my own car          25 - 45         cars   
21453     13127.587             to buy a car          25 - 45         cars   

      debt_profile   income_status  
0             paid  average earner  
1             paid      low earner  
2             paid      low earner  
3             paid  average earner  
4             paid      low earner  
...            ...             ...  
21449         paid      low earner  
21450         paid      low earner  
21451    defaulted      low earner  
21452    defaulted      low earner  
21453         paid      low earner  

[21454 rows x 16 columns]

**Categorizing the "children" column** 


In [87]:
def children_no(children):
#created function "children_no" with children as the argument
    
    if children == 0:
        return "no child"
    #if the number of children is 0,return the string "no child"
    elif children > 0:
        return "children"
     #if the number of children is greater than 0,return the string "children"
        

df["children_no"] = df["children"].apply(children_no)  
    



In [88]:
df["children"].value_counts()

0.0    14214
1.0     4808
2.0     2052
3.0      330
4.0       41
5.0        9
Name: children, dtype: int64

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [89]:
# Check the children data and paying back on time
df_children_data = df.pivot_table(index= "children_no", columns = "debt_profile", values = "dob_years", aggfunc = "count")
df_children_data


debt_profile  defaulted   paid
children_no                   
children            669   6571
no child           1072  13142

**Default rate for those with kids**

In [90]:
#Calculating the default percentage for those with kids
df[(df["debt_profile"]== "defaulted") & (df["children_no"]=="children")].shape[0]/df.loc[df["children_no"]=="children"].shape[0]

0.09240331491712707

**Default rate for those without kids**

In [91]:
#Calculating the default percentage for those without kids
df[(df["debt_profile"]== "defaulted") & (df["children_no"]=="no child")].shape[0]/df.loc[df["children_no"]=="no child"].shape[0]

0.07541860137892219

**Conclusion**

From the findings above, customers with kids are more likely to default on their loan than customers without kids 

**Is there a correlation between family status and paying back on time?**

The "family_status" column is grouped to five distinct categories, namely: civil partnership, divorced, married, unmarried, widow/widower. So, I grouped by the family_status table, and count for those who made payments on time, and those who defaulted. Then I calculated the default rate for each category

In [92]:
# Check the family status data and paying back on time
df_family_status = df.pivot_table(index= "family_status", columns = "debt_profile", values = "dob_years", aggfunc = "count")
df_family_status


debt_profile       defaulted   paid
family_status                      
civil partnership        388   3763
divorced                  85   1110
married                  931  11408
unmarried                274   2536
widow / widower           63    896

**Civil partnership default rate**

In [93]:
#Calculating the default rate for civil partnership
df[(df["debt_profile"]== "defaulted") & (df["family_status"]=="civil partnership")].shape[0]/df.loc[df["family_status"]=="civil partnership"].shape[0]

0.09347145266200915

**Divorced default rate**

In [94]:
#Calculating the default rate for those that are divorced
df[(df["debt_profile"]== "defaulted") & (df["family_status"]=="divorced")].shape[0]/df.loc[df["family_status"]=="divorced"].shape[0]

0.07112970711297072

**Married default rate**

In [95]:
#Calculating the default rate for widow/widower
df[(df["debt_profile"]== "defaulted") & (df["family_status"]=="married")].shape[0]/df.loc[df["family_status"]=="married"].shape[0]

0.07545181943431396

**Unmarried default rate**

In [96]:
#Calculating the default rate for unmarried people
df[(df["debt_profile"]== "defaulted") & (df["family_status"]=="unmarried")].shape[0]/df.loc[df["family_status"]=="unmarried"].shape[0]

0.09750889679715302

**Default rate for widow/widower**


In [97]:
df[(df["debt_profile"]== "defaulted") & (df["family_status"]=="widow / widower")].shape[0]/df.loc[df["family_status"]=="widow / widower"].shape[0]

0.06569343065693431

**Conclusion**


The highest default rate belongs to the "unmarried" and "civil partnership" groups 9.3% and 9.7% respectively, next are the "married" and "divorced"  which are 7.5% and 7.1%. The lowest default rate are from the "widows/widowed" which is 6.6%. 

**Is there a correlation between income level and paying back on time?**

The "total_income" column was grouped to six distinct categories, namely: Extremely High earner, High earner, Very High earner, above average earner, average earner, low earner, and stored in the "income_status" column. So, I grouped by the income_status table, and count for those who made payments on time, and those who defaulted. Then I calculated the default rate for each category

In [98]:
# Check the income level data and paying back on time

df_income_level = df.pivot_table(index= "income_status", columns = "debt_profile", values = "dob_years", aggfunc = "count")
df_income_level

debt_profile           defaulted     paid
income_status                            
Extremely High earner        1.0      5.0
High earner                  3.0     31.0
Very High earner             NaN      9.0
above average earner        10.0    163.0
average earner             180.0   2411.0
low earner                1547.0  17094.0

In [99]:
# Calculating default rate for Extremely High earners
df[(df["debt_profile"]== "defaulted") & (df["income_status"]== "Extremely High earner")].shape[0]/df.loc[df["income_status"]=="Extremely High earner"].shape[0]

0.16666666666666666

In [100]:
# Calculating default rate for High earners
df[(df["debt_profile"]== "defaulted") & (df["income_status"]== "High earner")].shape[0]/df.loc[df["income_status"]=="High earner"].shape[0]

0.08823529411764706

In [101]:
# Calculating default rate for above average earners
df[(df["debt_profile"]== "defaulted") & (df["income_status"]== "above average earner")].shape[0]/df.loc[df["income_status"]=="above average earner"].shape[0]

0.057803468208092484

In [102]:
# Calculating default rate average earners
df[(df["debt_profile"]== "defaulted") & (df["income_status"]== "average earner")].shape[0]/df.loc[df["income_status"]=="average earner"].shape[0]

0.06947124662292552

In [103]:
# Calculating default rate for low earners
df[(df["debt_profile"]== "defaulted") & (df["income_status"]== "low earner")].shape[0]/df.loc[df["income_status"]=="low earner"].shape[0]


0.08298911002628614

**Conclusion**

"Low earners" have a high probability of defaulting on their loans. While earners that are "average earners"  and above to those that are "Extremely High earners" are highly unlikely to default on their loans

**How does credit purpose affect the default rate?**

The purpose column was rearranged to four distinct values, namely: cars, real estate, wedding, education, and stored into purpose_norm. So, I grouped by the purpose_norm table, and count for those who made payments on time, and those who defaulted. Then I calculate the default rate for each category.

In [104]:
# Check the purpose_norm data and paying back on time

df_credit_purpose = df.pivot_table(index= "purpose_norm", columns = "debt_profile", values = "dob_years", aggfunc = "count")
df_credit_purpose

debt_profile  defaulted   paid
purpose_norm                  
cars                403   3903
education           370   3643
real estate         782  10029
wedding             186   2138

**Default rate for Car Purchase Loans**

In [105]:
# Checking the percentage default rate for Car Purchase

df[(df["purpose_norm"]== "cars") & (df["debt_profile"]== "defaulted")].shape[0]/df.loc[df["purpose_norm"]=="cars"].shape[0]

0.09359033906177427

**Default rate for Education**

In [106]:
# Checking the percentage default rate for Education

df[(df["purpose_norm"]== "education") & (df["debt_profile"]== "defaulted")].shape[0]/df.loc[df["purpose_norm"]=="education"].shape[0]

0.0922003488661849

**Default rate for Real Estate**

In [107]:
# Checking the percentage default rate for Real Estate

df[(df["purpose_norm"]== "real estate") & (df["debt_profile"]== "defaulted")].shape[0]/df.loc[df["purpose_norm"]=="real estate"].shape[0]

0.0723337341596522

**Default rate for wedding**

In [108]:
# Checking the percentage default rate for Wedding

df[(df["purpose_norm"]== "wedding") & (df["debt_profile"]== "defaulted")].shape[0]/df.loc[df["purpose_norm"]=="wedding"].shape[0]

0.08003442340791739

**Conclusion**

The default rate for car loans and education are the highest (above 9%), next is wedding loans, just 8%. The least default rate is Real Estate which is just above 7%


# General Conclusion 


In this Project I worked with a dataset that contained information on customers' income, loan purpose, debt status, family status, and children status. I had missing values that were symmetric on the total income, and the days employed. These missing values were not filled missing the global mean or global median. I identified customers' characteristics that gave more information about the distribution of the mean, and the median. 
The total income column was filled with the median, while the day employed column was filled with the mean. Duplicate values in the whole dataset were taken out. Days employed had two issues. It had negative values, and also it had outliers in terms of days that a person worked. First, I took the absolute value of the whole column, then I filtered out ages of people that were older than 65 years and replaced it with the median value of the whole column.
The children column had negative values of -1 and ouliers of -20. Although they were small, those values were not taken out, they were rather replaced with the  median.

We created dictionaries for family status, and education. The main dataset was grouped into purpose, children number, income status, debt profile and dob years. We found the default rates for each category in each of those columns. We found that:

Default rates for car loans and education loans are the highest (above 9%), followed by wedding loans (about 8%), and the lowest being real estate loans (just above 7%).
Low earners" have a high probabilitty of defaulting on their loans. While earners that are "average earners" and above to those that are "Extremely High earners" are highly unlikely to default on their loans.
The highest default rates with respect to family status belong to the "unmarried" and "civil partnership" groups (9.7% and 9.3%), followed by the married and divorced groups (7.5% and 7.1%), the lowest default rate belongs to the "widowed" group (6.6%).
Customers with kids are more likely(9.2%) to default on their loan (that is, not pay on time) than customers without kids(7.5%). From the analysis of this data, the Bank will find the categories with lower default rates more credit worthy i.e No kids, High Income, widowed, and taking a loan for real estate. Least favorable will be those from high default rate categories i.e having kids, low income, unmarried or in a civil partnership, and taking a car or education loan.




